<a href="https://colab.research.google.com/github/ghimirebimal/ML-Projects/blob/main/Basic_Deep_Learning_with_Neural_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##### Copyright  The TensorFlow Authors.

In [ ]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

##Basic Deep Learning with Neural Networks

Like every first app we should start with something simple that shows the overall scaffolding for how your code works. Let us consider one example that shows relationship between two numbers. So, for example, if you were writing code for a function like this, you already know the 'rules'  

```
float hw_function(float x){
    float y = (3 * x) - 1;
    return y;
}
```

So how would you train a neural network to do the equivalent task? Using data! By feeding it with a set of Xs, and a set of Ys, it should be able to figure out the relationship between them. 

This is obviously a very different paradigm than what you might be used to, so let's step through it piece by piece.

```
Rules' are the outcome of learning model based in input (data) and output(answer)
```


## Imports

Let's start with our imports. Here we are importing TensorFlow and calling it tf for ease of use.  We then import a library called numpy, which helps us to represent our data as lists easily and quickly. The framework for defining a neural network as a set of Sequential layers is called keras, so we import that too.

(**Note:** You can run the notebook using TensorFlow 2.5.0)

In [ ]:
#!pip install tensorflow==2.5.0

In [ ]:
import tensorflow as tf
import numpy as np
import pandas as pd
from tensorflow import keras
print(tf.__version__)

2.6.0


## Define and Compile the Neural Network

Next we will create the simplest possible neural network. It has 1 layer, and that layer has 1 neuron, and the input shape to it is just 1 value.

In [ ]:
model = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])


Now we compile our Neural Network. When we do so, we have to specify 2 functions, a loss and an optimizer.

If you've seen lots of math for machine learning, here's where it's usually used, but in this case it's nicely encapsulated in functions for you. But what happens here — let's explain...

We know that in our function, the relationship between the numbers is y=3x-1. 

When the computer is trying to 'learn' that, it makes a guess...maybe y=10x+10. The LOSS function measures the guessed answers against the known correct answers and measures how well or how badly it did.

It then uses the OPTIMIZER function to make another guess. Based on how the loss function went, it will try to minimize the loss. At that point maybe it will come up with somehting like y=5x+5, which, while still pretty bad, is closer to the correct result (i.e. the loss is lower) 

Note: Stochastic Gradient Descent (SGD) is a simple yet very efficient approach to fitting linear classifiers and regressors under convex loss functions such as (linear) Support Vector Machines and Logistic Regression.


It will repeat this for the number of EPOCHS which you will see shortly. But first, here's how we tell it to use 'MEAN SQUARED ERROR' for the loss and 'STOCHASTIC GRADIENT DESCENT' for the optimizer. You don't need to understand the math for these yet, but you can see that they work! :)

Over time you will learn the different and appropriate loss and optimizer functions for different scenarios. 


In [ ]:
model.compile(optimizer='sgd', loss='mean_squared_error')


## Providing the Data

Next up we'll feed in some data. In this case we are taking 6 xs and 6ys. You can see that the relationship between these is that y=3x-1, so where x = -2, y=-7 etc. etc. 

A python library called 'Numpy' provides lots of array type data structures that are a defacto standard way of doing it. We declare that we want to use these by specifying the values as an np.array[]

In [ ]:
xs = np.array([-2.0,	-1.0,	0.0, 1.0,	2.0,	3.0], dtype=float)
ys = np.array([-7.0,	-4.0,	-1.0,	2.0,	5.0, 8.0], dtype=float)


In [ ]:
model.fit(xs,ys,epochs=500)

# Training the Neural Network

The process of training the neural network, where it 'learns' the relationship between the Xs and Ys is in the **model.fit**  call. This is where it will go through the loop we spoke about above, making a guess, measuring how good or bad it is (aka the loss), using the opimizer to make another guess etc. It will do it for the number of epochs you specify. When you run this code, you'll see the loss on the right hand side.

Ok, now you have a model that has been trained to learn the relationship between X and Y. You can use the **model.predict** method to have it figure out the Y for a previously unknown X. So, for example, if X = 11, what do you think Y will be? Take a guess before you run this code:

In [ ]:
print(model.predict([11.0]))
#y=3*X-1=32

[[31.999704]]


You might have thought 32, right? But it ended up being a little under. Why do you think that is? 

Remember that neural networks deal with probabilities, so given the data that we fed the NN with, it calculated that there is a very high probability that the relationship between X and Y is Y=3X-1, but with only 6 data points we can't know for sure. As a result, the result for 11 is very close to 32, but not necessarily 32. 

As you work with neural networks, you'll see this pattern recurring. You will almost always deal with probabilities, not certainties, and will do a little bit of coding to figure out what the result is based on the probabilities, particularly when it comes to classification.


**(10 points) Write a simple program that predicts**
* the house price where the base price is $300,000.00 and price increases by \$100k per bedroom. Predict the price for (say) 6 bedroom house. **(5 points)**

In [ ]:
def hw_function(x):
     y = (300000 + x*100000)
     return (y)


In [ ]:
for x in range(1, 6):  
   print(hw_function(x))

400000
500000
600000
700000
800000


In [ ]:
xh = np.array([1,	2,	3,	4,	5], dtype=float)
yh = np.array([400000,	500000,	600000,	700000,	800000], dtype=float)
yh=yh/1000

In [ ]:
model_house = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

In [ ]:
model_house.compile(optimizer='sgd', loss='mean_squared_error')


In [ ]:
model_house.fit(xh, yh, epochs=1500)


In [ ]:
print(1000*model_house.predict([6]))

[[901041.3]]


Initially, when data was applied without scaling, the loss was increasing excessively. So, the target value is scaled by 1000. Trained Neural Network model predicted $901041.3 for 6 bedroom house when the number of epochs is choosen 1500. At 500 epochs, it was predicting around $930K. 





*   the credit limit where the base credit limit is \$500.00 and credit limit increases by \$500 per \$5000 increment in annual income. Predict the credit limit for (say) \$100,000.00 annual salary. **(5 points)**



For the given xc and yc values below, with successive epochs, the loss (mean square error) was raising enormosly and reached to infinity. So, instead of origional data, z score normalization is used to address increasing loss value.

In [ ]:
xc=np.array([0,5000,10000,15000,20000,25000],dtype=float)
yc=np.array([500,1000,1500,2000,2500,3000], dtype=float)
xc_mean=np.mean(xc)
xc_std=np.std(xc)
xc_zscore=np.array(list((map(lambda x:(x-xc_mean)/xc_std,xc))))

In [ ]:
model_credit = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])


In [ ]:
model_credit.compile(optimizer='sgd', loss='mean_squared_error')

In [ ]:
model_credit.fit(xc_zscore, yc, epochs=1000)

In [ ]:
#Annual income
inc=100000
inc_zscore=(inc-xc_mean)/xc_std
print(model_credit.predict([inc_zscore]))

[[10499.982]]
